feat: handle encryption (and therefore message delivery) to invalid addresses

nventuro opened the pull request:
Closes https://linear.app/aztec-labs/issue/F-188/handle-invalid-addresses.

Instead of skipping log emission, I decided to encrypt to a random point. This is better as it results in zero additional constraints.

benesjan commented on noir-projects/aztec-nr/aztec/src/messages/encryption/aes128.nr:
```diff
@@ -174,10 +174,28 @@ impl MessageEncryption for AES128 {
+                .to_address_point()
+                .unwrap_or(
+                    // Safety: if the recipient is an invalid address, then it is not possible to encrypt a message for
+                    // them because we cannot stablish a shared secret. This is never expected to occur during normal
```
```suggestion
                    // them because we cannot establish a shared secret. This is never expected to occur during normal
```

benesjan commented on noir-projects/aztec-nr/aztec/src/messages/encryption/aes128.nr:
```diff
@@ -174,10 +174,28 @@ impl MessageEncryption for AES128 {
 
         // (not to be confused with the tagging shared secret)
-        // TODO (#17158): Currently we unwrap the Option returned by derive_ecdh_shared_secret_using_aztec_address.
+        // TODO (#17158): Currently we unwrap the Option returned by derive_ecdh_shared_secret.
```
```suggestion
```

benesjan commented on noir-projects/aztec-nr/aztec/src/messages/encryption/aes128.nr:
```diff
@@ -384,15 +402,34 @@ impl MessageEncryption for AES128 {
     }
 }
 
+/// Produces a random valid address point, i.e. one that is in the curve. This is equivalent to calling
```
```suggestion
/// Produces a random valid address point, i.e. one that is on the curve. This is equivalent to calling
```

benesjan commented on noir-projects/aztec-nr/aztec/src/messages/encryption/aes128.nr:
```diff
@@ -384,15 +402,34 @@ impl MessageEncryption for AES128 {
+    let mut result = std::mem::zeroed();
+
+    loop {
+        // We simply produce random x coordinates until we find one that is in the curve. About half of the x
```
```suggestion
        // We simply produce random x coordinates until we find one that is on the curve. About half of the x
```

benesjan reviewed (approved):
Very nice!

The comment mentioning "king on the hill" attack is great.

mverzilli commented on noir-projects/aztec-nr/aztec/src/keys/ecdh_shared_secret.nr:
```diff
@@ -3,40 +3,19 @@ use dep::protocol_types::{
-/// # Formula
-/// `[ephemeral_secret] * recipient_address_public_key = shared_secret`
+/// The input secret is known only to one party. The output shared secret can be derived given knowledge of
+/// `public_key`'s key-pair and the public ephemeral secret, using this same function (with reversed inpus).
```
```suggestion
/// `public_key`'s key-pair and the public ephemeral secret, using this same function (with reversed inputs).
```

mverzilli reviewed (approved):
🚀

nventuro committed:
feat: handle encryption (and therefore message delivery) to invalid addresses

Closes https://linear.app/aztec-labs/issue/F-188/handle-invalid-addresses.

Instead of skipping log emission, I decided to encrypt to a random point. This is better as it results in zero additional constraints.
